src/backend/distributed/metadata/node_metadata.c

```diff
@@ -1131,16 +837,93 @@ PrimaryNodeForGroup(int32 groupId, bool *groupContainsNodes)
+	 * we activate the nodes inside metadataSyncContext.
+	 * 1) sets isactive to false, which means no query is allowed from coordinator
+	 *    to the nodes,
+	 * 2) sets metadatasynced to false, which means coordinator will not sync DDL
+	 *    to the nodes.
+	 *
+	 * NOTE: We use separate connection to localhost as we will rollback local
+	 * transaction. We only deactivate nodes at coordinator. Operator should
+	 * handle problems at workers if any. (workers can see each other as active and
+	 * can try to send queries. In that case, proper metadata sync for the workers
+	 * should be done.)
+	 */
+
+	int connectionFlag = FORCE_NEW_CONNECTION;
+	MultiConnection *connection = GetNodeConnection(connectionFlag, LocalHostName,
+													PostPortNumber);
+
+	bool activeAndSynced = false;
+	WorkerNode *workerNode = NULL;
+	foreach_ptr(workerNode, context->activatedWorkerNodeList)
+	{
+		char *deactivateCommand = NodeStateUpdateCommand(workerNode->nodeId,
+														 activeAndSynced);
+		char *desyncCommand = NodeHasmetadataUpdateCommand(workerNode->nodeId,
+														   activeAndSynced);
+		List *commandList = list_make2(deactivateCommand, desyncCommand);
+		SendCommandListToWorkerOutsideTransactionWithConnection(connection, commandList);
+	}
+
+	CloseConnection(connection);
+}
+
+
+/*
+ * ActivateNodeList does some sanity checks and acquire Exclusive lock on pg_dist_node,
+ * and then iterates over the nodeList and activates the nodes.
+ *
+ * The function operates in 3 different modes according to transactionMode inside
+ * metadataSyncContext.
+ *
+ * 1. MetadataSyncCollectsCommands(context):
+ *      Only collect commands instead of sending them to workers,
+ * 2. context.transactionMode == METADATA_SYNC_TRANSACTIONAL:
+ *      Send all commands using coordinated transaction,
+ * 3. context.transactionMode == METADATA_SYNC_NON_TRANSACTIONAL:
+ *      Send all commands using bare (no transaction block) connections.
  */
 void
-ActivateNodeList(List *nodeList)
+ActivateNodeList(MetadataSyncContext *context)
 {
+	if (context->activatedWorkerNodeList == NIL)
+	{
+		/*
+		 * in case user calls with only coordinator in nodelist, we can hit here. Just bail
+		 * out as we already warned user, at `SetMetadataSyncNodesFromNodeList`, that
+		 * coordinator has metadata.
+		 */
+		return;
+	}
+
+	if (context->transactionMode == METADATA_SYNC_NON_TRANSACTIONAL &&
+		IsMultiStatementTransaction())
+	{
+		/*
+		 * prevent inside transaction block as we use bare connections which can
+		 * lead deadlock
+		 */
+		ereport(ERROR, (errmsg("citus does not allow sync metadata inside "
+							   "a transaction block when sync mode is nontransactional"),
+						errhint("resync after SET citus.metadata_sync_transaction_mode "
+								"TO 'transactional'")));
+	}
+
 	/*
-	 * We currently require the object propagation to happen via superuser,
+	 * we currently require the object propagation to happen via superuser,
```

Comment:
I think all multi-line comments should start with capital. In your patch, you do the other way around. Do you have a specific reason for that?